File: apps/owner-portal/lib/apiClient.ts
const API_BASE = process.env.NEXT_PUBLIC_API_BASE_URL || "https://backend-8qpa.onrender.com/api";

export function getToken() {
  if (typeof window === "undefined") return null;
  return localStorage.getItem("authToken");
}

export function clearAuthAndRedirect() {
  if (typeof window !== "undefined") {
    localStorage.removeItem("authToken");
    window.location.href = "/login";
  }
}

export async function apiFetch(path: string, options: RequestInit = {}) {
  const token = getToken();

  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    ...(options.headers as Record<string, string>),
  };

  if (token) {
    headers["Authorization"] = `Bearer ${token}`;
  }

  const res = await fetch(`${API_BASE}${path}`, {
    ...options,
    headers,
    cache: "no-store", // Force no cache for admin dashboard responses
  });

  if (res.status === 401) {
    clearAuthAndRedirect();
    throw new Error("Session expired. Please log in again.");
  }

  const text = await res.text();
  const data = text ? JSON.parse(text) : {};

  if (!res.ok) {
    if (res.status === 403) {
      if (typeof window !== "undefined") window.location.href = "/not-allowed";
    }
    throw new Error(data.message || `Request failed with status ${res.status}`);
  }

  return data;
}

export const api = {
  get: (path: string, options?: RequestInit) => apiFetch(path, { ...options, method: "GET" }),
  post: (path: string, body?: any, options?: RequestInit) =>
    apiFetch(path, { ...options, method: "POST", body: JSON.stringify(body) }),
  patch: (path: string, body?: any, options?: RequestInit) =>
    apiFetch(path, { ...options, method: "PATCH", body: JSON.stringify(body) }),
  delete: (path: string, options?: RequestInit) => apiFetch(path, { ...options, method: "DELETE" }),
};

File: apps/delivery_partner_app/lib/core/network/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../config/env.dart';

class ApiClient {
  ApiClient._internal();
  static final ApiClient instance = ApiClient._internal();

  final http.Client _client = http.Client();

  Future<String?> _getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('auth_token');
  }

  Future<Map<String, String>> _defaultHeaders({bool withAuth = true}) async {
    final headers = <String, String>{
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };

    if (withAuth) {
      final token = await _getToken();
      if (token != null && token.isNotEmpty) {
        headers['Authorization'] = 'Bearer $token';
      }
    }

    return headers;
  }

  Uri _buildUri(String path) {
    return Uri.parse('${Env.apiBaseUrl}$path');
  }

  Future<http.Response> post(
    String path, {
    Map<String, dynamic>? body,
    bool withAuth = true,
  }) async {
    final headers = await _defaultHeaders(withAuth: withAuth);
    final uri = _buildUri(path);
    return _client.post(
      uri,
      headers: headers,
      body: jsonEncode(body ?? {}),
    );
  }

  Future<http.Response> get(
    String path, {
    bool withAuth = true,
  }) async {
    final headers = await _defaultHeaders(withAuth: withAuth);
    final uri = _buildUri(path);
    return _client.get(uri, headers: headers);
  }

  Future<http.Response> patch(
    String path, {
    Map<String, dynamic>? body,
    bool withAuth = true,
  }) async {
    final headers = await _defaultHeaders(withAuth: withAuth);
    final uri = _buildUri(path);
    return _client.patch(
      uri,
      headers: headers,
      body: jsonEncode(body ?? {}),
    );
  }
}